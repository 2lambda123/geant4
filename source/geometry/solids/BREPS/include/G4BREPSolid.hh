#ifndef __SOLID_H
#define __SOLID_H
#include "G4VSolid.hh"	      
#include "G4VisExtent.hh"      
#include "G4Surface.hh"
#include "G4Axis2Placement3D.hh"
#include "G4PointRat.hh"	 
#include "G4BoundingBox3D.hh"	 

class STEPentity;
class InstMgr;
class G4Ray;


class G4BREPSolid : public G4VSolid
{

public:

  G4BREPSolid(const G4String name);
  G4BREPSolid(const G4String, G4Surface**, G4int);
  ~G4BREPSolid();

  virtual G4String GetEntityType() const {return "Closed_Shell";}  
  virtual void Initialize();
  G4int CreateSTEPData(); // not yet implemented 

  G4bool CalculateExtent(const EAxis              pAxis      ,
			 const G4VoxelLimits&     pVoxelLimit,
			 const G4AffineTransform& pTransform ,
			 G4double&                pMin       , 
			 G4double&                pMax        ) const;

  virtual EInside Inside(register const G4ThreeVector&) const;

  virtual G4ThreeVector SurfaceNormal(const G4ThreeVector&) const;

  virtual G4double DistanceToIn(const G4ThreeVector&) const;
  virtual G4double DistanceToIn(register const G4ThreeVector&,
				register const G4ThreeVector&) const;

  virtual G4double DistanceToOut(const G4ThreeVector&) const;
  virtual G4double DistanceToOut(register const G4ThreeVector&,
				 register const G4ThreeVector&,
				 const G4bool  calcNorm=false , 
				 G4bool        *validNorm=0   , 
				 G4ThreeVector *n=0             ) const;
 

  G4Point3D Scope(); // ???

  void DescribeYourselfTo (G4VGraphicsScene& scene) const;
  G4VisExtent   GetExtent        () const;
  G4Polyhedron* CreatePolyhedron () const;
  G4NURBS*      CreateNURBS      () const;

  G4int Intersect(register const G4Ray&)const;

  inline G4double IntersectionDistance()const{return intersectionDistance;}
  void IntersectionDistance(const G4double d)const 
  {
    ((G4BREPSolid*)this)->intersectionDistance=d;
  }   

  G4Surface* GetSurface(G4int nr)
  {
    return SurfaceVec[nr];
  }
  
  inline void Active(const G4int x)const 
  {
    ((G4BREPSolid*)this)->active=x;
  }
  
  inline G4int Active() const {return active;}
  
  virtual inline void Reset() const
  {
    ((G4BREPSolid*)this)->active=1;
    ((G4BREPSolid*)this)->intersectionDistance=kInfinity;
    ((G4BREPSolid*)this)->startInside=0; 
     
    for(register G4int a=0;a<nb_of_surfaces;a++)
      SurfaceVec[a]->Reset();

    ShortestDistance = kInfinity;
  }

  static G4int NumberOfSolids;
  static InstMgr InstanceList;

  G4double GetShortestDistance() const {return ShortestDistance;}

  G4int GetId() const {return Id;}

  void SetId(G4int id) {Id = id;}

  G4String GetName() const {return solidname;}

  void SetName(G4String name) {solidname = name;}

  G4int NumberOfFaces() const {return nb_of_surfaces;} 

  // Add by L. Broglia
  G4Axis2Placement3D* GetPlace() { return place; }
  G4BoundingBox3D*    GetBBox()  { return bbox;  } 
 
protected:

  G4bool  IsConvex();
  virtual void CalcBBoxes();
  void    CheckSurfaceNormals();
  void    RemoveHiddenFaces(register const G4Ray& G4Rayref, G4int)const;   
  void    TestSurfaceBBoxes(register const G4Ray&) const;

  inline G4int StartInside() const 
  {
    return startInside;
  }
  
  inline void StartInside(const G4int si) const 
  {
    ((G4BREPSolid*)this)->startInside=si;
  }  


private:

  G4int IsBox();
  G4int FinalEvaluation(register const G4Ray&, const G4int =0) const;


protected:
  G4Axis2Placement3D* place;
  static G4Ray        Track;
  static G4double     ShortestDistance;
  G4int               Box, Convex, AxisBox, PlaneSolid;
  G4BoundingBox3D*    bbox;   
  G4double            intersectionDistance;
  G4int               active;
  G4int               startInside;
  G4int               nb_of_surfaces;
  G4Point3D           intersection_point;
  G4Surface**         SurfaceVec;
  G4double            RealDist;
  G4String            solidname; 
  G4int               Id;
   

  void QuickSort( register G4Surface** SrfVec, 
		  register G4int left, register G4int right) const
  {
    register G4int i=left;
    register G4int j=right;
    register G4Surface* elem1;
    register G4Surface* elem2 = SrfVec[(left+right)/2];
    register G4double tmpdistance;
    do
    {
      tmpdistance = elem2->Distance();
      while ( SrfVec[i]->Distance() < tmpdistance  && i < right ) i++;
      while (tmpdistance < SrfVec[j]->Distance()  && j > left ) j--;
      if(i<=j)
      {
	elem1 = SrfVec[i];
	SrfVec[i] = SrfVec[j];
	SrfVec[j] = elem1;
	i++;j--;
      }
     } while (i<=j);
      
     if( left < j  ) QuickSort(SrfVec,left, j );
     if( i < right ) QuickSort(SrfVec,i, right);    
    }

};

#endif









