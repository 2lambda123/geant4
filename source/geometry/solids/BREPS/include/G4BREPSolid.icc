// This code implementation is the intellectual property of
// the GEANT4 collaboration.
//
// By copying, distributing or modifying the Program (or any work
// based on the Program) you indicate your acceptance of this statement,
// and all its terms.
//
// $Id: G4BREPSolid.icc,v 1.3 2000-11-08 14:21:58 gcosmo Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// --------------------------------------------------------------------
// GEANT 4 inline definitions file
//
// G4BREPSolid.icc
//
// Implementation of inline methods of G4BREPSolid
// --------------------------------------------------------------------

inline G4double G4BREPSolid::IntersectionDistance() const
{
  return intersectionDistance;
}

inline void G4BREPSolid::IntersectionDistance(G4double d) const 
{
  ((G4BREPSolid*)this)->intersectionDistance=d;
}   

inline G4Surface* G4BREPSolid::GetSurface(G4int nr) const
{
  return SurfaceVec[nr];
}
  
inline void G4BREPSolid::Active(G4int x) const 
{
  ((G4BREPSolid*)this)->active=x;
}
  
inline G4int G4BREPSolid::Active() const
{
  return active;
}

inline G4double G4BREPSolid::GetShortestDistance() const
{
  return ShortestDistance;
}

inline G4int G4BREPSolid::GetId() const
{
  return Id;
}

inline void G4BREPSolid::SetId(G4int n)
{
  Id = n;
}

inline const G4String& G4BREPSolid::GetName() const
{
  return solidname;
}

inline void G4BREPSolid::SetName(const G4String& name)
{
  solidname = name;
}

inline G4int G4BREPSolid::GetNumberOfSolids() const
{
  return NumberOfSolids;
}

inline G4int G4BREPSolid::GetNumberOfFaces() const
{
  return nb_of_surfaces;
}

inline const G4Axis2Placement3D* G4BREPSolid::GetPlace() const
{
  return place;
}

inline const G4BoundingBox3D* G4BREPSolid::GetBBox() const
{
  return bbox;
} 

inline G4int G4BREPSolid::StartInside() const 
{
  return startInside;
}
  
inline void G4BREPSolid::StartInside(G4int si) const 
{
  ((G4BREPSolid*)this)->startInside=si;
}  

inline void G4BREPSolid::QuickSort( register G4Surface** SrfVec, 
		                    register G4int left,
				    register G4int right) const
{
    register G4int i=left;
    register G4int j=right;
    register G4Surface* elem1;
    register G4Surface* elem2 = SrfVec[(left+right)/2];
    register G4double tmpdistance;
    do
    {
      tmpdistance = elem2->GetDistance();
      while ( SrfVec[i]->GetDistance() < tmpdistance  && i < right ) i++;
      while (tmpdistance < SrfVec[j]->GetDistance()  && j > left ) j--;
      if(i<=j)
      {
	elem1 = SrfVec[i];
	SrfVec[i] = SrfVec[j];
	SrfVec[j] = elem1;
	i++;j--;
      }
    } while (i<=j);
      
    if( left < j  ) QuickSort(SrfVec,left, j );
    if( i < right ) QuickSort(SrfVec,i, right);    
}
