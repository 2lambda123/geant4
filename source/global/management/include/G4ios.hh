// This code implementation is the intellectual property of
// the GEANT4 collaboration.
//
// By copying, distributing or modifying the Program (or any work
// based on the Program) you indicate your acceptance of this statement,
// and all its terms.
//
// $Id: G4ios.hh,v 1.4 1999-11-23 15:00:04 gcosmo Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// 
// ---------------------------------------------------------------
// GEANT 4 class header file
//
// G4ios.hh
//
// ---------------------------------------------------------------
#ifndef included_G4ios
#define included_G4ios

#if defined(OO_DDL_TRANSLATION)
/*
 * stdlib needs to be included before iostream.h
 * during oddlx runs to work around a parser
 * problem of AIX ooddlx v4.0.2 with some versions of
 * AIX system header files.
 */
#include <stdlib.h>
#endif

#include "g4std/iostream"
#include "G4Types.hh"

extern G4std::ostream G4cout;
extern G4std::ostream G4cerr;

#define G4cin G4std::cin
#define G4endl G4std::endl

#endif
