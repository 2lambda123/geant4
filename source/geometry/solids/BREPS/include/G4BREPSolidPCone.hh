//
// ********************************************************************
// * DISCLAIMER                                                       *
// *                                                                  *
// * The following disclaimer summarizes all the specific disclaimers *
// * of contributors to this software. The specific disclaimers,which *
// * govern, are listed with their locations in:                      *
// *   http://cern.ch/geant4/license                                  *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.                                                             *
// *                                                                  *
// * This  code  implementation is the  intellectual property  of the *
// * GEANT4 collaboration.                                            *
// * By copying,  distributing  or modifying the Program (or any work *
// * based  on  the Program)  you indicate  your  acceptance of  this *
// * statement, and all its terms.                                    *
// ********************************************************************
//
//
// $Id: G4BREPSolidPCone.hh,v 1.9 2002-01-22 22:43:01 radoone Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//
// ----------------------------------------------------------------------
// Class G4BREPSolidPCone
//
// Class description:
//
//  The polyconical solid G4BREPSolidPCone is a shape defined by a set of 
//  inner and outer conical or cylindrical surface sections and two planes 
//  perpendicular to the Z axis. Each conical surface is defined by its 
//  radius at two different planes perpendicular to the Z-axis. Inner and 
//  outer conical surfaces are defined using common Z planes:
//
//  G4BREPSolidPCone( const G4String& name,
//                          G4double start_angle,
//                          G4double opening_angle,		   
//                          G4int    num_z_planes,
//                          G4double z_start,		   
//                          G4double z_values[],
//                          G4double RMIN[],
//                          G4double RMAX[] )

// Authors: J.Sulkimo, P.Urban.
// Revisions by: L.Broglia, S.Giani, G.Cosmo.
// ----------------------------------------------------------------------
#ifndef __G4BREPSolidPCone
#define __G4BREPSolidPCone

#include "G4BREPSolid.hh"

class G4BREPSolidPCone : public G4BREPSolid
{

 public: // with description

  G4BREPSolidPCone( const G4String& name,
		    G4double start_angle,
		    G4double opening_angle,		   
		    G4int    num_z_planes, // sections
		    G4double z_start,		   
		    G4double z_values[],
		    G4double RMIN[],
		    G4double RMAX[] );
    // Constructor defining the polycone through its
    // conical/cylindrical surfaces.

  ~G4BREPSolidPCone();
    // Destructor.

  void Initialize();
    // Calculates the bounding box.

  EInside Inside(register const G4ThreeVector& Pt) const;
    // Determines if the point Pt is inside, outside or on the surface
    // of the solid.

  G4ThreeVector SurfaceNormal(const G4ThreeVector&) const;
    // Calculates the normal of the surface at a point on the surface
    // The sense of the normal depends on the sense of the surface.

  G4double DistanceToIn(const G4ThreeVector&) const;
    // Calculates the shortest distance ("safety") from a point
    // outside the solid to any boundary of this solid.
    // Return 0 if the point is already inside.

  G4double DistanceToIn(register const G4ThreeVector& Pt, 
			register const G4ThreeVector& V) const;
    // Calculates the distance from a point Pt outside the solid
    // to the solid's boundary along a specified direction vector V.
    // Note: Intersections with boundaries less than the tolerance must
    //       be ignored if the direction is away from the boundary.

  G4double DistanceToOut(register const G4ThreeVector& Pt, 
			 register const G4ThreeVector& V, 
			 const G4bool calcNorm=false, 
			 G4bool *validNorm=0, G4ThreeVector *n=0) const;
    // Calculates the distance from a point inside the solid to the solid`s
    // boundary along a specified direction vector.
    // Return 0 if the point is already outside.
    // Note: If the shortest distance to a boundary is less than the
    //       tolerance, it is ignored. This allows for a point within a
    //       tolerant boundary to leave immediately.

  G4double DistanceToOut(const G4ThreeVector&) const;
    // Calculates the shortest distance ("safety") from a point inside the
    // solid to any boundary of this solid.
    // Return 0 if the point is already outside.	

public:

  void Reset() const;
    // Resets all distance attributes.

  G4Polyhedron* CreatePolyhedron () const;
    // Creates a G4Polyhedron

private:
  
  G4Surface* ComputePlanarSurface( G4double r1, G4double r2,
                                   G4ThreeVector& origin,
                                   G4ThreeVector& planeAxis,
                                   G4ThreeVector& planeDirection,
                                   G4int surfSense );
    // For a given radius values compute a planar surface

  G4BREPSolidPCone(const G4BREPSolidPCone&);
  G4BREPSolidPCone& operator=(const G4BREPSolidPCone&);
    // Private copy constructor and assignment operator.

  // The following is only utilised in storing the shape parameters for
  // use in visualising this shape.  J.A. Feb  24, 1997
  //
  struct PConeParameters {
     G4double Start_angle;
     G4double Opening_angle;		   
     G4int    Num_z_planes; 
     // G4double z_start;		   
     G4double *Z_values;
     G4double *Rmin;
     G4double *Rmax;
  }  original_parameters;
};

#endif
