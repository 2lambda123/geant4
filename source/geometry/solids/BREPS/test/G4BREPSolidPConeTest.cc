//
// ********************************************************************
// * DISCLAIMER                                                       *
// *                                                                  *
// * The following disclaimer summarizes all the specific disclaimers *
// * of contributors to this software. The specific disclaimers,which *
// * govern, are listed with their locations in:                      *
// *   http://cern.ch/geant4/license                                  *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.                                                             *
// *                                                                  *
// * This  code  implementation is the  intellectual property  of the *
// * GEANT4 collaboration.                                            *
// * By copying,  distributing  or modifying the Program (or any work *
// * based  on  the Program)  you indicate  your  acceptance of  this *
// * statement, and all its terms.                                    *
// ********************************************************************
//
//////////////////////////////////////////////////////////////////////////
// $Id: G4BREPSolidPConeTest.cc,v 1.12 2002-01-28 16:16:23 radoone Exp $
// GEANT4 tag $Name: not supported by cvs2svn $
//////////////////////////////////////////////////////////////////////////
//
//
// BREP solid test, create by L. Broglia, 20/10/98
// modification of old G4Gerep test
//


#include "G4Timer.hh"
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include "G4ios.hh" 
#include "G4BREPSolid.hh"
#include "G4BREPSolidPCone.hh"

#include "g4std/fstream"
#include "g4std/iomanip"

void checkNormal( G4BREPSolid* solid, G4ThreeVector& position )
{
  G4ThreeVector normal = solid->SurfaceNormal( position );
  
  G4cout << G4endl
         << "\t\tSurface normal"                 << G4endl
         << "\t\t--------------"                 << G4endl
         << "\t\tposition x="                    << position.x()
         << " y="                                << position.y()
         << " z="                                << position.z() << G4endl
         << "\t\tis normX="                      << normal.x()
         << " normY="                            << normal.y()
         << " normZ="                            << normal.z()   << G4endl;
}

void checkSurfInOut( G4BREPSolid* solid, G4ThreeVector& position, G4ThreeVector& direction )
{
  G4double d;
  
  G4cout << G4endl
         << "\tTest of In & Out " << solid->GetName()   << G4endl
         << "\t--------------------------------------" << G4endl;
  
  d = solid->DistanceToIn( position, direction );
  G4cout << "\tDistance to in="          << d;
  d = solid->DistanceToIn( position );
  G4cout << "\tClosest distance to in="  << d          <<G4endl;
    
  d = solid->DistanceToOut( position, direction );
  G4cout << "\tDistance to out="         << d;
  d = solid->DistanceToOut( position );
  G4cout << "\tClosest distance to out=" << d          << G4endl;
}

void checkSolid( const G4String& where, G4BREPSolid* solid, G4ThreeVector& position, G4ThreeVector& direction, G4double estIn = 0.0, G4double estOut = 0.0 )
{
  G4double d;
  
  EInside in = solid->Inside( position );
  
  G4cout << G4endl
         << where << " of " << solid->GetName()                   << G4endl
         << "--------------------------------" << G4endl
         << "estimation to in  =" << estIn     << G4endl
         << "estimation to out =" << estOut    << G4endl
         << "direction x=" << direction.x()
         << "  y=" << direction.y()
         << "  z=" << direction.z()            << G4endl
         << "position x=" << position.x()
         << "  y=" << position.y()
         << "  z=" << position.z();

  if( in == kInside ) {
    G4cout <<" is inside";
    d = solid->DistanceToOut( position, direction );
    G4cout<<"  distance to out="<<d;
    d = solid->DistanceToOut( position );
    G4cout<<"  closest distance to out="<<d<<G4endl;
  } else if( in == kOutside ) {
    G4cout <<" is outside";
    d = solid->DistanceToIn( position, direction );
    G4cout<<"  distance to in="<<d;
    d = solid->DistanceToIn( position );
    G4cout<<"  closest distance to in="<<d<<G4endl;
  } else {
    G4cout <<" is on the surface"<<G4endl;
    checkSurfInOut( solid, position, direction );
    checkNormal( solid, position );
  }
}

int main(G4int argc, char **argv)
{
  const G4int noZplanes= 8; 
  
  G4double RMINVec[noZplanes]  = {  30,  30,   0,  0,   0,   0,  40,  40 };
  G4double RMAXVec[noZplanes]  = {  70,  70,  70, 40,  40,  80,  80,  60 };
  G4double Z_Values[noZplanes] = { -20, -10, -10,  0,  10,  20,  30,  40 };

  G4double start_angle= 0.0;
  G4double opening_angle= 2*pi;

  G4double zstart= Z_Values[0]; 

  G4cout << "\n=======     PCon test      ========";

  G4BREPSolidPCone *MyPCone = new G4BREPSolidPCone ("MyPCone",
						    start_angle,
						    opening_angle,
						    noZplanes,
						    zstart,
						    Z_Values,
						    RMINVec,
						    RMAXVec );
  
  G4cout << "\n\nPCone created ! "<<G4endl;
  G4cout << "Variety is G4BREPSolidPolycone"<<G4endl;

  G4cout << "Its parameters are: "<<G4endl;

  ///////////////////////////////////////////////////
  // Temporary
  for (G4int x = 0; x < noZplanes; x++)
  {
    G4cout <<    " Z[" << x << "]=" << G4std::setw(5) << Z_Values[x];
    G4cout << " Rmin[" << x << "]=" << G4std::setw(5) << RMINVec[x];
    G4cout << " Rmax[" << x << "]=" << G4std::setw(5) << RMAXVec[x]<<G4endl;
  }

  G4cout<<" start   angle ="<<start_angle<<G4endl;
  G4cout<<" opening angle ="<<opening_angle<<G4endl;
  G4cout<<" zstart =" << zstart << G4endl;


  // -> Check methods :
  //  - Inside
  //  - DistanceToIn
  //  - DistanceToOut

  
  EInside in;
  
  G4cout<<"\n\n==================================================";
  G4ThreeVector  pt(0, -100, 24);
  G4double y; 
  for (y = -100; y<=100; y+=10)
  {
    pt.setY(y);
    in = MyPCone->Inside(pt);
    
    G4cout << "\nx=" << pt.x() << "  y=" << pt.y() << "  z=" << pt.z();
    
    if( in == kInside )
      G4cout <<" is inside";
    else
      if( in == kOutside )
	G4cout <<" is outside";
      else
	G4cout <<" is on the surface";
  }

  G4cout<<"\n\n==================================================";
  G4ThreeVector  start( 0, 0, -30);
  G4ThreeVector dir(1, 1, 0);
  G4double   d;
  
  G4cout<<"\nPdep is (0, 0, z)";
  G4cout<<"\nDir is (1, 1, 0)\n";

  G4double z;
  for(z=-30; z<=50; z+=5)
  {
    start.setZ(z);

    in = MyPCone->Inside(start);
    G4cout<< "x=" << start.x() << "  y=" << start.y() << "  z=" << start.z();
    
    if( in == kInside )
    {
      G4cout <<" is inside";

      d = MyPCone->DistanceToOut(start, dir);
      G4cout<<"  distance to out="<<d;
      d = MyPCone->DistanceToOut(start);
      G4cout<<"  closest distance to out="<<d<<G4endl;
    }
    else if( in == kOutside ) 
    {
      G4cout <<" is outside";

      d = MyPCone->DistanceToIn(start, dir);
      G4cout<<"  distance to in="<<d;
      d = MyPCone->DistanceToIn(start);
      G4cout<<"  closest distance to in="<<d<<G4endl;
    }
    else
      G4cout <<" is on the surface"<<G4endl;

  }

  G4cout<<"\n\n==================================================";
  G4ThreeVector  start2( 0, -100, -30);
  G4ThreeVector dir2(0, 1, 0);
  G4double   d2;

  G4cout<<"\nPdep is (0, -100, z)";
  G4cout<<"\nDir is (0, 1, 0)\n";

  for(z=-30; z<=50; z+=5)
  {
    G4cout<<"  z="<<z;
    start2.setZ(z);
    d2 = MyPCone->DistanceToIn(start2, dir2);
    G4cout<<"  distance to in="<<d2;
    d2 = MyPCone->DistanceToIn(start2);
    G4cout<<"  closest distance to in="<<d2<<G4endl;
  }

  G4cout<<"\n\n==================================================";
  G4ThreeVector  start3( 0, 0, -50);
  G4ThreeVector dir3(0, 0, 1);
  G4double   d3;

  G4cout<<"\nPdep is (0, y, -50)";
  G4cout<<"\nDir is (0, 0, 1)\n";

  for(y=-0; y<=90; y+=5)
  {
    G4cout<<"  y="<<y;
    start3.setY(y);
    d3 = MyPCone->DistanceToIn(start3, dir3);
    G4cout<<"  distance to in="<<d3<<G4endl;
  }
  
  ///////////////////////////////////////////////////////////////////////////////////////
  // Test due to the bugfix No. 320
  // Solid setup (only profile of its upper half is shown):
  //
  //     +----+
  //  +--+    +--+
  //  +--+    +--+
  //     +----+
  //
  // -------------- Z axis
  //
 	G4double zValTOB[6]     = { -2800*mm, -1210*mm, -1210*mm, 1210*mm, 1210*mm, 2800*mm };
 	G4double rMinTOB[6]     = { 1160*mm,   1160*mm,  540*mm,  540*mm,  1160*mm, 1160*mm };
 	G4double rMaxTOB[6]     = { 1170*mm,   1170*mm,  1182*mm, 1182*mm, 1170*mm, 1170*mm };
 	G4BREPSolidPCone * cmsTOBSolid = new G4BREPSolidPCone( "BREPPolyconeTOBSolid",
                                                         0.0*rad, twopi*rad,
                                                  			 6, zValTOB[0], zValTOB, rMinTOB, 
                                                         rMaxTOB );
  // The critical gun settings
  G4ThreeVector gunPosition( 904.05833958335438*mm, -761.44764667689935*mm, -382.29360686839397*mm );
  G4ThreeVector gunDirection( -0.32132043849994285, 0.11159991009963287, 0.94037154139624957 );

  checkSolid( "CMS case", cmsTOBSolid, gunPosition, gunDirection );  

  G4ThreeVector gp( 541*mm, 0*mm, 1210*mm );
  G4ThreeVector gd( -0.32132043849994285, 0.11159991009963287, 0.94037154139624957 );
  
  checkSolid( "surface", cmsTOBSolid, gp, gd );  
  
  G4ThreeVector gpedge( 540*mm, 540*mm, -1210*mm );
  G4ThreeVector gdedge( -0.32132043849994285, 0.11159991009963287, 0.94037154139624957 );
  
  checkSolid( "edge", cmsTOBSolid, gpedge, gdedge );  
  
  G4ThreeVector gpin( 540*mm, 540*mm, -1209*mm );
  G4ThreeVector gdin( -0.32132043849994285, 0.11159991009963287, 0.94037154139624957 );
  
  checkSolid( "inside", cmsTOBSolid, gpin, gdin );  

 	G4double zVal1[6]     = { -200*mm,  -100*mm, -100*mm, 100*mm, 100*mm, 200*mm };
 	G4double rMin1[6]     = {   50*mm,    50*mm,   25*mm,  25*mm,  50*mm,  50*mm };
// 	G4double rMin1[6]     = {   50*mm,    50*mm,   50*mm,  50*mm,  50*mm,  50*mm };
 	G4double rMax1[6]     = {   50*mm,   100*mm,  200*mm, 200*mm, 100*mm,  50*mm };
 	G4BREPSolidPCone* s1  = new G4BREPSolidPCone( "s1", 0.0*rad, twopi*rad, 6, zVal1[0], zVal1, rMin1, rMax1 );
    
  G4ThreeVector gps1in( 150*mm, 0*mm, 0*mm );
  G4ThreeVector gds1in( 0, 0, 1 );
  G4ThreeVector gps1out( 0*mm, 0*mm, 0*mm );
  G4ThreeVector gds1out( 1, 0, 0 );
  G4ThreeVector gps1edge( 50*mm, 0*mm, -200*mm );
  G4ThreeVector gds1edge( 0, 0, 1 );
  G4ThreeVector gps1surfout( 200*mm, 0*mm, 0*mm );
  G4ThreeVector gds1surfout( -1, 0, 0 );
  G4ThreeVector gps1surfin( 0*mm, 25*mm, 0*mm );
  G4ThreeVector gds1surfin( 0, 1, 0 );
  G4ThreeVector gps1plansurfleftout( 150*mm, 0*mm, -100*mm );
  G4ThreeVector gps1plansurfleftin( 30*mm, 0*mm, -100*mm );
  G4ThreeVector gps1plansurfleftfict( 0*mm, 0*mm, -100*mm );
  G4ThreeVector gds1plansurfleft( 0, 0, 1 );
  G4ThreeVector gps1plansurfright( 150*mm, 0*mm, 100*mm );
  G4ThreeVector gds1plansurfright( 0, 0, -1 );
  
  checkSolid( "inside", s1, gps1in, gds1in );  
  checkSolid( "outside", s1, gps1out, gds1out );  
  checkSolid( "edge", s1, gps1edge, gds1edge, 0, 400 );  
  checkSolid( "outer surface", s1, gps1surfout, gds1surfout, 0, 100 );  
  checkSolid( "inner surface", s1, gps1surfin, gds1surfin, 0, 175 );  
  checkSolid( "left outer planar surface", s1, gps1plansurfleftout, gds1plansurfleft, 0, 200 );  
  checkSolid( "left outer planar surface inv", s1, gps1plansurfleftout, gds1plansurfright, kInfinity, 0 );  
  checkSolid( "fictious left planar surface", s1, gps1plansurfleftfict, gds1plansurfleft, kInfinity, 25 );  
  checkSolid( "fictious left planar surface inv", s1, gps1plansurfleftfict, gds1plansurfright, kInfinity, 25 );  
  checkSolid( "left inner planar surface", s1, gps1plansurfleftin, gds1plansurfleft, 0, 200 );  
  checkSolid( "left inner planar surface inv", s1, gps1plansurfleftin, gds1plansurfright, kInfinity, 0 );  
  checkSolid( "right planar surface", s1, gps1plansurfright, gds1plansurfright, 0, 200 );  
  checkSolid( "right planar surface inv", s1, gps1plansurfright, gds1plansurfleft, 0 );  

  return EXIT_SUCCESS;
}

