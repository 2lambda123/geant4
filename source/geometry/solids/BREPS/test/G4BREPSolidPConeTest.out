
=======     PCon test      ========

PCone created ! 


==================================================
x=0  y=-100  z=24 is outside
x=0  y=-90  z=24 is outside
x=0  y=-80  z=24 is outside
x=0  y=-70  z=24 is inside
x=0  y=-60  z=24 is inside
x=0  y=-50  z=24 is inside
x=0  y=-40  z=24 is inside
x=0  y=-30  z=24 is inside
x=0  y=-20  z=24 is inside
x=0  y=-10  z=24 is outside
x=0  y=0  z=24 is outside
x=0  y=10  z=24 is outside
x=0  y=20  z=24 is inside
x=0  y=30  z=24 is inside
x=0  y=40  z=24 is inside
x=0  y=50  z=24 is inside
x=0  y=60  z=24 is inside
x=0  y=70  z=24 is inside
x=0  y=80  z=24 is outside
x=0  y=90  z=24 is outside
x=0  y=100  z=24 is outside

==================================================
Pdep is (0, 0, z)
Dir is (1, 1, 0)
  z=-30  distance to in=9e+99
  z=-25  distance to in=9e+99
  z=-20  distance to in=30
  z=-15  distance to in=30
  z=-10  distance to in=30
  z=-5  distance to in=55
  z=0  distance to in=40
  z=5  distance to in=40
  z=10  distance to in=40
  z=15  distance to in=60
  z=20  distance to in=80
  z=25  distance to in=20
  z=30  distance to in=40
  z=35  distance to in=40
  z=40  distance to in=40
  z=45  distance to in=9e+99
  z=50  distance to in=9e+99


==================================================
Pdep is (0, -100, z)
Dir is (0, 1, 0)
  z=-30  distance to in=9e+99
  z=-25  distance to in=9e+99
  z=-20  distance to in=30
  z=-15  distance to in=30
  z=-10  distance to in=30
  z=-5  distance to in=45
  z=0  distance to in=60
  z=5  distance to in=60
  z=10  distance to in=60
  z=15  distance to in=40
  z=20  distance to in=20
  z=25  distance to in=20
  z=30  distance to in=20
  z=35  distance to in=30
  z=40  distance to in=40
  z=45  distance to in=9e+99
  z=50  distance to in=9e+99


==================================================
Pdep is (0, y, -50)
Dir is (0, 0, 1)
  y=0  distance to in=50
  y=5  distance to in=48.75
  y=10  distance to in=47.5
  y=15  distance to in=46.25
  y=20  distance to in=45
  y=25  distance to in=43.75
  y=30  distance to in=42.5
  y=35  distance to in=41.25
  y=40  distance to in=40
  y=45  distance to in=38.75
  y=50  distance to in=37.5
  y=55  distance to in=36.25
  y=60  distance to in=35
  y=65  distance to in=33.75
  y=70  distance to in=32.5
  y=75  distance to in=31.25
  y=80  distance to in=30
  y=85  distance to in=9e+99
  y=90  distance to in=9e+99
